// @sp-api-sdk/auth/dist/es/src/access-token.js
import { AxiosError } from 'axios';
import { SellingPartnerApiAuthError } from './error';
import { axios } from './utils/axios';
export class AccessTokenFactory {
    clientId;
    clientSecret;
    refreshToken;
    scopes;
    value;
    expirationDate;
    constructor(parameters) {
        this.clientId = parameters.clientId;
        this.clientSecret = parameters.clientSecret;
        this.refreshToken = parameters.refreshToken;
        this.scopes = parameters.scopes;
    }
    async get() {
        if (!this.value || (this.expirationDate && Date.now() >= this.expirationDate.getTime())) {
            const body = {
                client_id: this.clientId,
                client_secret: this.clientSecret,
                ...(this.refreshToken
                    ? {
                        grant_type: 'refresh_token',
                        refresh_token: this.refreshToken,
                    }
                    : {
                        grant_type: 'client_credentials',
                        scope: this.scopes.join(' '),
                    }),
            };
            try {
                const { data } = await axios.post('/o2/token', body);
                const expiration = new Date();
                expiration.setSeconds(expiration.getSeconds() + data.expires_in);
                this.value = data;
                this.expirationDate = expiration;
            }
            catch (error) {
                if (error instanceof AxiosError) {
                    throw new SellingPartnerApiAuthError(error);
                }
                throw error;
            }
        }
        return this.value.access_token;
    }
}
